WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainStateHolder.kt

```diff
@@ -16,161 +16,215 @@
 package com.example.android.wearable.speaker
 
 import android.Manifest
-import android.app.Application
+import android.app.Activity
 import android.content.pm.PackageManager
 import android.media.AudioDeviceInfo
 import android.media.AudioManager
-import androidx.constraintlayout.motion.widget.MotionLayout
+import android.media.MediaPlayer
+import androidx.annotation.RequiresPermission
+import androidx.compose.foundation.MutatorMutex
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
 import androidx.core.content.ContextCompat
 import androidx.core.content.getSystemService
-import kotlinx.coroutines.flow.MutableStateFlow
-import kotlinx.coroutines.flow.StateFlow
-import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import java.time.Duration
+import kotlin.coroutines.resume
 
 /**
- * A state holder driving the logic of [MainActivity].
+ * A state holder driving the logic of the app.
+ *
+ * This state holder is scoped to the composition (and therefore can be provided an [Activity]) to handle
+ * permission-related logic.
  */
 class MainStateHolder(
-    private val application: Application,
+    private val activity: Activity,
     private val requestPermission: () -> Unit,
     private val showPermissionRationale: () -> Unit,
     private val showSpeakerNotSupported: () -> Unit
 ) {
-
     /**
-     * The backing [MutableStateFlow] for the [appState].
+     * The [MutatorMutex] that guards the state of the app.
+     *
+     * Due to being a [MutatorMutex], this automatically handles cleanup of any ongoing asynchronous work,
+     * like playing music or recording, ensuring that only one operation is occurring at a time.
      */
-    private val appStateMutableStateFlow = MutableStateFlow<AppState>(AppState.Ready(transitionInstantly = true))
+    private val appStateMutatorMutex = MutatorMutex()
```

Comment:
@manuelvicnt @JoseAlcerreca I'd like to get your input on this AppState implementation.

Reply:
The usage of the `MutatorMutex` looks ok but I wonder if it would be simpler to just have the `appState` backed by a `StateFlow` and use the `update` function that gives you atomic updates out of the box.

I'm not sure if this will be the way of atomically setting values for `mutableStateOf`s in the future as you have also transactions in Compose.

Reply:
The `MutatorMutex` here is doing more work than just atomically updating `mutableStateOf`:

Since I'm expressing all of the recording and playback as `suspend` functions, I'm using the `MutatorMutex` to ensure that only one thing is happening at a time.

So if the user taps on the "Play music" button, and then taps on it again halfway through the song, the second tap will interrupt the playback due to the `appStateMutatorMutex. mutate` call, and the old playback will be cleaned up using the normal coroutine cancellation mechanisms.

An alternative solution here would be keeping track of an ongoing `Job` directly, but that solution ends up looking like a `MutatorMutex` anyway.

Reply:
SGTM, thanks for explaining! Let's add more documentation around that as comments :D it's a very important behavior to keep in mind